### PR TITLE
Improve popup position

### PR DESCRIPTION
**Pull request by karthago1**
Make the popup positions more consistent.
Improvements:
1. if the signature popup content is bigger than the available space, then the popup is always shown under the cursor, even if there more space above the cursor than below. [check this video](https://asciinema.org/a/JNF85TAwa2bsZJWOHW70dstJU) 
2. There is no mutation anymore inside required_size. Maybe in the future we can update all widgets to have no mutations and change the trait

the following is the new behavior
[asciicast](https://asciinema.org/a/44ZvwHoJX7UGZiPfLpnZR8asY)

IMO this make at least the completion dialog consistent and will be shown always bellow the cursor.

Edit: Video Updated

### CODE REVIEW

File: helix-term/src/ui/popup.rs
Comment (pascalkuthe) [inline helix-term/src/ui/popup.rs:279]:
```suggestion
            .required_size((area.width, area.height))
```

File: helix-term/src/ui/popup.rs
Comment (pascalkuthe) [inline helix-term/src/ui/popup.rs:265]: If you look at thedDefinition of required size it says you should not be clamping the size to the size of the viewport. I would like to follow the documented definition of the function. The reason the old code was buggy and confusing is because people did adhox things that kind of worked but didn't follow the actual api.

File: helix-term/src/ui/menu.rs
Comment (pascalkuthe) [inline helix-term/src/ui/menu.rs:397]: May be missing something but why was this removed?

File: helix-term/src/ui/popup.rs
Comment (pascalkuthe) [inline helix-term/src/ui/popup.rs:127]: I think this should stay as `get_ir_insert` otherwise the popus move with the cursor. I also have some larger changes for this in #6417 IIRC since cursor calulation changes/is made more efficient there. 

I also wouldn't want to recompile teh cursor position all the time (which is slow)

File: helix-term/src/ui/popup.rs
Comment (pascalkuthe) [inline helix-term/src/ui/popup.rs:142]: Why are there always 4 lines below the cursor? The statusline and promp would be just two (also add that to this comment, if it's not clear to me it's not clear to other readers)

**Review by pascalkuthe — COMMENTED** [review]
I think this looks pretty promising. I left some comments but I didn't have time to read trough everything I detail yet.

File: helix-term/src/ui/popup.rs
Comment (karthago1) [reply helix-term/src/ui/popup.rs]: Thank you for bugfix

File: helix-term/src/ui/popup.rs
Comment (karthago1) [reply helix-term/src/ui/popup.rs:264]: I think I understood the API very well. What this code is doing is that, it reads the content requested_size, which can  be infinite and clamp it to the popup maximum size, which IMO fine, because the popup shouldn't request higher size than it's maximum size

This PR didn't change anything regarding the requested size of the popup.  It only removes the mutation of `self.size` and `self.child_size` at the cost that requested_size will be called twice in the completion popup, for the other popup there should be no drawbacks.

if you don't like removing the mutation, maybe I can modify the PR to only improve the popup behavior without much modifications

File: helix-term/src/ui/menu.rs
Comment (karthago1) [reply helix-term/src/ui/menu.rs]: sorry it was a mistake. (this why it was in draft, because it was late)

File: helix-term/src/ui/popup.rs
Comment (karthago1) [reply helix-term/src/ui/popup.rs]: I tried to remove the mutation of `area` function. but missed that. already reverted

File: helix-term/src/ui/popup.rs
Comment (karthago1) [reply helix-term/src/ui/popup.rs:141]: Right now I am assuming that the popup has a minimum size of 4. (we can make it configurable). only if this minimum Size is not met, we change the location of the popup. This actually what this PR about. IMO this improves the popup location much. (Right now, there is some cases where the popup changes the location without important reason)

In that comment, I said, since the minimum size of the popup can be 4 and below the cursor I have always at least 6 free lines, where the popup can be drawn (see screenshot below). So I think that the flag can_put_below is useless. 
popup

nevertheless I think we should keep that flag. maybe in the future we can make the minimum popup size configurable and some popups will request higher minimum size. Than this flag will be important.
(I remove that comment for the time being)

Comment (pascalkuthe) [reply helix-term/src/ui/popup.rs:141]: hmm yeah almost everything your are saying is what I got from the code too just the following part:

> below the cursor I have always at least 6 free lines, where the popup can be drawn (see screenshot below)

this is not true. I think you are talking about scrolloff here but that can be configured/disabled (`:set scrolloff 0` so that assumption does not hold in general. 

So I would just remove that comment.

File: helix-term/src/ui/popup.rs
Comment (pascalkuthe) [reply helix-term/src/ui/popup.rs:264]: I quite like removing the mutation.  I think you got confused where I put the review comment (at the top of the function). I was talking about the following code:

```
            (width + self.margin.width()).min(max_width),
            (height + self.margin.height()).min(max_height),
```

The capping (`.min`) is what shouldn't happen in required size. That is the API contract. Instead you should do that wen calling `required_size` (I think in area)

**Comment by pascalkuthe** [comment]
Judging from your video there is either an existing bug or something new introduced with this PR going on. The signature help should be hidden automatically if it interests with the completions. In stead case the completion doc popup seem to overlap.

File: helix-term/src/ui/popup.rs
Comment (karthago1) [reply helix-term/src/ui/popup.rs:264]: the code you mentioned was already there. check the changes :sweat_smile: , this PR didn't change anything regarding the output of `required_size`

> The capping (.min) is what shouldn't happen in required size. That is the API contract. Instead you should do that wen calling required_size (I think in area)

I disagree for the following reason: the maximum size of the popup is hard coded to (120,26). why should the popup's `required_size` returns any higher value than its maximum size? IMO the current (and old) implementation comply with the API description, unless you want to remove that hard coded maximum size and allow the Popup to have any size it likes.

**Comment by karthago1** [comment]
yes I saw that, I will check it that issue and try to fix it

File: helix-term/src/ui/popup.rs
Comment (pascalkuthe) [reply helix-term/src/ui/popup.rs:261]: yes I understand that you didn't change this code but I am saying that the old code was wrong for doing this and you should fix it as part of this PR.

On second thaught I think the misunderstanding is coming from the fact that `required_size` just doens't make sense for popup it all. `required_size` is an *optinal* method of component so that containers can compute its size and add a scrollbar.

I was saying you shouldn't limit the size of the popup to the viewport here since that isn't the job of required_size (you want to return the true size so the parent add a scrollbar. Particularly I took issue with limiting it to the viewport size since the viewport size would be the size of the entire window, the true size computation depends on the cursor position and can't be done here. T

You are right that the width/height should be limited to the maximum popup width/height (120/26). I was not taking issue with that but the limiting to the viewport size which can messup calculations in area later (or is just useless most of the time).

I think we should just not implement `required_size` for popup at all. `required_size` really only make sense for components that go *inside a popup` and take up their whole parent (and are therefore cropped to the viewport). Instead, the logic should just be moved into the `area` function directly (that function does its own cropping anyway I think then it will be more clear why I took issue with that)

Comment (karthago1) [reply helix-term/src/ui/popup.rs:261]: sorry it took me a while to understand the issue :smiley: . basically those 2 lines does not follow the API description:

```
let max_width = 120.min(viewport.0);
let max_height = 26.min(viewport.1.saturating_sub(2)); // add some spacing in the viewport
```
you are right. Somehow I missed this. I think, this can be moved to the `area` function without any side effects or issues. What I don't like about the API is `mut self`. also in the `area` function there is `mut self`. but In this PR I am not going to make such huge changes to the API. I will keep it for another one :smile_cat: (If I find a better solution)

Comment (karthago1) [reply helix-term/src/ui/popup.rs:264]: I hope this solved/fine

**Comment by karthago1** [comment]
Until now the `SignatureHelp` `Popup` is removed and triggered in the `Completion` logic. I removed this logic and instead the completion view store it's used areas in the Editor. Later when rendering the `SignatureHelp`. the completion areas (from the editor) are checked and if they intersect, then it is not rendered.

the current solution is more responsive when comes to hide/show the popups.

(I have update the video in the PR description)

File: helix-term/src/handlers/signature_help.rs
Comment (pascalkuthe) [inline helix-term/src/handlers/signature_help.rs:233]: I do prefer the `Above` bias for signature help. Why was that moved underneath the cursor? Particularly with autcompletions now being almost always below that just seems like the natural choice.

**Comment by pascalkuthe** [comment]
> Until now the `SignatureHelp` `Popup` is removed and triggered in the `Completion` logic. I removed this logic and instead the completion view store it's used areas in the Editor. Later when rendering the `SignatureHelp`. the completion areas (from the editor) are checked and if they intersect, then it is not rendered.
> 
> the current solution is more responsive when comes to hide/show the popups.
> 
> (I have update the video in the PR description)

hmm I don't love that. I get why you implemented it that way, there is no other way to share this information and it may change at any point due to resizing. I am not a fan of that tough. I want to really stop adding ad-hoc fields to `Editor`. I also want to move the component infrastructure to be a more typical passive/immutable widget tree with no mutation inside the rendering (except maybe for some caches but that can be interior mutability). 

For example one possibility could be that we use either somekind of popup container that manages all open popus to hide overlapping popups (or do it in the compositor directly that's really where that kind of thing belongs IMO). 

There could also be other approaches I have some ideas but I think all of them will involve larger changes. I don't think you need to tackle that here. Ther rest of the PR is a nice not too large change so I would prefer to just revert the changes you made to the signature help popup and rather merge this as is. Sorry for not thinking about that earlier when I made that comment I thaught that was maybe a new regression with this PR.

File: helix-term/src/handlers/signature_help.rs
Comment (karthago1) [reply helix-term/src/handlers/signature_help.rs]: It is still above. the `SignaturePopup` is new struct it will set itself Above in the `new` method. but this is not important anymore, since this is reverted

**Comment by karthago1** [comment]
I reverted as you asked.

> I want to really stop adding ad-hoc fields to `Editor`. 

we can put the field directly under `Context`. `Context` is usually used for bad stuff :smile: 

> For example one possibility could be that we use either somekind of popup container that manages all open popus to hide overlapping popups (or do it in the compositor directly that's really where that kind of thing belongs IMO).

I thought about a `PopupManager` but then the adhoc was too attractive and easy to implement and IMO it leaves the code base in better state, since it remove any dependency from the `Completion` to `Signature` Popup.

> The first step for that is turning the completions menu into a proper component instead of the spaghetti code we currently have (that is a historical artifact that I think is not necessary anymore now with the event system). I am already working on that in a much larger PR that makes manly changes to the completion menu (the refactor there started for unrelated reasons but I see now that its really the right change anyway)

nice then I will wait for your PR first before making further changes

>  when I made that comment I thaught that was maybe a new regression with this PR.

yes it is not. Th regression appears only if there no space except in top, and the window must be very small. So I think this PR is still improving the popup positions much. (In my screen I never encounter it.)

I am think of creating a new PR to add some key binding to toggle on/off the Signature Popup. Sometimes it is annoying.  DO you know if it is possible to toggle off the Signature popup in helix?

**Review by pascalkuthe — DISMISSED** [review]

**Comment by karthago1** [comment]
@pascalkuthe there is bug, which also exists in the master branch.

`.expect("Component needs required_size implemented in order to be embedded in a popup");`

this panic actually can happen, if the popup is open and console area get smaller. (I got this somehow while dragging the console.) I have already fixed this. I prefer to keep the changes as small as possible to get MRs merged quickly, (especially there is an approve now :smile: ) should I add some commits to this MR or better wait until this get merged?

**Comment by pascalkuthe** [comment]
> @pascalkuthe there is bug, which also exists in the master branch.
> 
> `.expect("Component needs required_size implemented in order to be embedded in a popup");`
> 
> this panic actually can happen, if the popup is open and console area get smaller. (I got this somehow while dragging the console.) I have already fixed this. I prefer to keep the changes as small as possible to get MRs merged quickly, (especially there is an approve now 😄 ) should I add some commits to this MR or better wait until this get merged?

can you push them to a seperate branch and link them? It should really be a bug in whatever component is embedded inside the popup rather than popup itself.

**Comment by karthago1** [comment]
done #10507. it shows much changes, because this commit is included there.

**Comment by karthago1** [comment]
as request move the commit 5deff632e72723c9f238e1b081a1f4ba69987832 to this PR. I didn't modify the other one

**Review by pascalkuthe — APPROVED** [review]

**Review by the-mikedavis — APPROVED** [review]

**Pull request merged**
